src/Portability/Converter.php

```diff
@@ -0,0 +1,246 @@
+                }
+
+                return $value;
+            };
+        }
+
+        if ($rightTrimString) {
+            $functions[] = static function ($value) {
+                if (! is_string($value)) {
+                    return $value;
+                }
+
+                return rtrim($value);
+            };
+        }
+
+        return $this->compose(...$functions);
+    }
+
+    /**
+     * Creates a function that will convert each array-row retrieved from the database
+     *
+     * @param callable|null $function The function that will convert each value
+     * @param int|null      $case     Column name case
+     *
+     * @return callable|null The resulting function or NULL if no conversion is needed
+     */
+    private function createConvertRow(?callable $function, ?int $case) : ?callable
+    {
+        $functions = [];
+
+        if ($function !== null) {
+            $functions[] = $this->createMapper($function);
+        }
+
+        if ($case !== null) {
+            $functions[] = static function (array $row) use ($case) : array {
+                return array_change_key_case($row, $case);
+            };
+        }
+
+        return $this->compose(...$functions);
+    }
+
+    /**
+     * Creates a function that will be applied to the return value of Statement::fetch*()
+     * or an identity function if no conversion is needed
+     *
+     * @param callable|null $function The function that will convert each tow
+     * @param callable      $id       Identity function
+     */
+    private function createConvert(?callable $function, callable $id) : callable
+    {
+        if ($function === null) {
+            return $id;
+        }
+
+        return static function ($value) use ($function) {
+            if ($value === false) {
+                return false;
+            }
+
+            return $function($value);
+        };
+    }
+
+    /**
+     * Creates a function that will be applied to the return value of Statement::fetchAll*()
+     * or an identity function if no transformation is required
+     *
+     * @param callable|null $function The function that will transform each value
+     * @param callable      $id       Identity function
+     */
+    private function createConvertAll(?callable $function, callable $id) : callable
+    {
+        if ($function === null) {
+            return $id;
+        }
+
+        return $this->createMapper($function);
+    }
+
+    /**
+     * Creates a function that maps each value of the array using the given function
+     *
+     * @param callable $function The function that maps each value of the array
+     */
+    private function createMapper(callable $function) : callable
+    {
+        return static function (array $array) use ($function) : array {
+            return array_map($function, $array);
+        };
+    }
+
+    /**
+     * Creates a composition of the given set of functions
+     *
+     * @param callable ...$functions The functions to compose
```

Comment:
I think you are going to have to use https://psalm.dev/docs/annotating_code/type_syntax/callable_types/ to fix the Psalm issue below

Reply:
I ended up changing the signature of the `array_reduce()`'s `$callback` so that it can handle `$carry = null`, and now Psalm seems happy.